### DISCUSSION TITLE
Hardcoded Audience does not work

### DISCUSSION THREAD

**Discussion by hantsy** [keycloak/keycloak#30565] [Q&A]
I followed the guide to add an audience to the given client - **demo-client** (enabled **Service Account roles** in Authenction flow), I used global **Realm Roles** in my project.

* In *Client Scopes*, I create a new client scope named **demo-service**
* In **demo-service**, I created a Mapper from *by configuration*, add audience, and included in client **demo-client**, type is Default. 
* Switch to **demo-client**, in **Client Scopes** tab, I added **demo-service** scope, and type is Default.

But when generating access token by command.

```bash 
curl http://localhost:8000/realms/demo/protocol/openid-connect/token 
-v  
-u "demo-client:OQttJYpKRAvcSMOdYkZYv3Y6mZ8EBLMA" 
-d "grant_type=password"  
-d "username=demouser" 
-d "password=password"  
-d "scope=profile%20email%20demo-service"
``` 
And decoded the access token in the https://jwt.io page, the `aud` field is a UUID string, not the value of `demo-service` or the custom value in the audience mapper page.

I used the latest Keycloak 25 from a docker compose file. Check [docker compose file here](https://github.com/hantsy/spring-security-oauth2-resourceserver-example/blob/master/keycloak/docker-compose.yml).

**Comment by matt3o** [comment]
Same problem here, super frustrating. I think this should be a bug right? No matter which options I try I can't get it to work

**Comment by bonnm** [comment]
Could you open  this as a new issue here?
https://github.com/keycloak/keycloak/issues
I often use hardcoded audience mappers, but directly bound to clients, not to a scope. And it always worked (KC 25 not yet tested)

**Reply by a-florian-h** [reply]
I could not have a weblate installation to accept keycloak and the error was: InvalidAudienceError 

Indeed as @hantsy said 
> the aud field is a UUID string, not the value of `the client`

Unless harcoded ... which in my case is sufficient.  
Thanks !

**Reply by hantsy** [reply]
I found there is a way to display the custom value in the `aud` in the access token.

When creating the Mapper of *Client Scopes*(Audience).
1. Leave *Included Client Audience* blank. Although this field is marked as required, but can be saved with blank.
2. Fill *Included Custom Audience*, it will be shown in the token.

**Reply by rannox** [reply]
> I found there is a way to display the custom value in the `aud` in the access token.
> 
> When creating the Mapper of _Client Scopes_(Audience).
> 
>     1. Leave _Included Client Audience_ blank. Although this field is marked as required, but can be saved with blank.
> 
>     2. Fill _Included Custom Audience_, it will be shown in the token.

Thank you very much! After spending hours figuring out why the hardcoded audience didn't work, your post saved me!  :)
